### PR TITLE
Implementation of the Coinbase Rosetta API

### PR DESCRIPTION
**Pull request by cmullaparthi**
This PR adds the read only parts of the Rosetta API to the node

- x `/network/list`
- x `/network/options`
- x `/network/status`
- x `/account/balance`
-   `/account/coins` (only required for UTXO blockchains, so not needed for AE)
- x `/block`
- x `/block/transaction`
- x `/mempool`
- x `/mempool/transaction` (partially supported - balance events missing)
-   `/construction/combine`
-   `/construction/derive`
-   `/construction/hash`
-   `/construction/metadata`
-   `/construction/parse`
-   `/construction/payloads`
-   `/construction/preprocess`
-   `/construction/submit`

PR sponsored by the ACF

### CODE REVIEW

File: apps/aehttp/src/aehttp_app.erl [hunk 1]
```diff
@@ -12,6 +12,8 @@
 -define(DEFAULT_SWAGGER_EXTERNAL_LISTEN_ADDRESS, <<"0.0.0.0">>).
 -define(DEFAULT_SWAGGER_INTERNAL_PORT, 8143).
 -define(DEFAULT_SWAGGER_INTERNAL_LISTEN_ADDRESS, <<"127.0.0.1">>).
+-define(DEFAULT_ROSETTA_PORT, 8243).
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_app.erl:15]: This is an entirely different default from the one in the aeternity config schema, and the sys.config files.
In general, at least I tend to rely on the schema defaults alone, esp. given that we can now override those values using OS env variables.

File: apps/aehttp/src/aehttp_spec.erl [hunk 2]
```diff
@@ -19,11 +19,24 @@ json(SpecVsn) ->
     Json = jsx:prettify(jsx:encode(Yaml)),
     Json.
 
+yaml_to_json(Yaml0) when is_binary(Yaml0) ->
+    Yaml =
+        lists:foldl(
+            fun(Fun, Accum) -> Fun(Accum) end,
+            Yaml0,
+            []),
```
[the remark this answers is not in the document]
Comment (seanhinde) [reply apps/aehttp/src/aehttp_spec.erl:27]: Interesting, yes. This code isn't used. I suspect it was intended to support generation of rosetta_endpoints.erl, which doesn't work yet. WIP

File: apps/aeapi/src/aeapi.erl [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+            % BlockTxs = lists:foldl(
+            %     fun(MicroBlock, Acc) ->
+            %         BlockTxs = micro_block_txs(MicroBlock),
+            %         {ok, MBHash} =
+            %             aec_headers:hash_header(
+            %                 aec_blocks:to_header(MicroBlock)),
+            %         Ops = format_txs(BlockTxs, MBHash),
+            %         [Ops | Acc]
+            %     end,
+            % [], MicroBlocks),
```
Comment (uwiger) [inline apps/aeapi/src/aeapi.erl:368]:
```suggestion
```

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl [hunk 4]
```diff
@@ -0,0 +1,659 @@
...
+            case aec_db:find_tx_with_location(TxHashInternal) of
+                none ->
+                    throw(block_not_found);
+                {mempool, _} ->
+                    throw(block_not_found);
+                {BlockHash, STx} ->
+                    %% Tx was in this microblock. Check this microblock is in the
+                    %% Requested keyblock
+                    case aec_chain:get_block(BlockHash) of
+                        {ok, MBlock} ->
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_rosetta.erl:283]: Should be done within one activity.

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl [hunk 5]
```diff
@@ -0,0 +1,659 @@
...
+                throw(tx_not_found)
+        end,
+    case aec_db:find_tx_location(TxHashInternal) of
+        mempool ->
+            ok;
+        _ ->
+            throw(tx_not_found)
+    end,
+    case aec_db:find_signed_tx(TxHashInternal) of
+        {value, SignedTx} ->
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_rosetta.erl:370]: Should be done within one activity

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl [hunk 6]
```diff
@@ -0,0 +1,659 @@
...
+            case aec_blocks:height(TopBlock) of
+                0 ->
+                    TopBlock;
+                _ ->
+                    {ok, Prev} = aeapi:prev_key_block(TopBlock),
+                    Prev
+            end,
+        CurrentBlockIdentifier = format_block_identifier(CurrentBlock),
+        CurrentBlockTimestamp = aec_blocks:time_in_msecs(CurrentBlock),
+        GenesisBlockIdentifier = format_block_identifier(aec_chain:genesis_block()),
```
Comment (uwiger) [inline apps/aehttp/src/aehttp_dispatch_rosetta.erl:141]: Should be done within one activity

Comment (uwiger) [reply apps/aehttp/src/aehttp_dispatch_rosetta.erl:141]: Perhaps consider whether activity wrappers should be placed uniformly at the top level, or whether some should be in helper functions. The `ensure_activity/2` semantics allows for a combination, but also consider what should be dirty and where transactions are needed. As a point of implementation, an outer `transaction` will convert inner `async_dirty` activities to re-use the outer transaction context.

File: config/sys.config [hunk 7]
```diff
@@ -24,6 +24,10 @@
           {acceptors, 10},
           {port, 3113}
           ]},
+      {rosetta, [
+          {acceptors, 10},
+          {port, 3213}
```
[the remark this answers is not in the document]
Comment (seanhinde) [reply config/sys.config:29]: Merely following convention :) I have a future plan to re-vamp the config system which will take care of this

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl (hunk 6 above)
Comment (seanhinde) [reply apps/aehttp/src/aehttp_dispatch_rosetta.erl:141]: Nice suggestion. I'll look at this in the next phase of Rosetta

File: apps/aecore/src/aec_dry_run.erl [hunk 8]
```diff
@@ -96,20 +96,23 @@ dry_run_int([], _Trees, Env, _Opts, Acc) ->
...
     Stateless = proplists:get_value(stateless, TxOpts, false),
     Env1 = prepare_env(Env, TxOpts),
+    EventsEnabled = proplists:get_bool(tx_events, Opts),
     %% GH3283: Here we should collect and present the `internal_call_tx` events.
     %% This means expanding the return type, and breaking the api :scream_cat:.
     case aec_trees:apply_txs_on_state_trees([Tx], Trees, Env1, [strict, dont_verify_signature|Opts]) of
         {ok, [Tx], [], Trees1, Events} when Stateless ->
-            Env2 = aetx_env:set_events(Env, Events),
-            dry_run_int(Txs, Trees, Env2, Opts, [dry_run_res(Tx, Trees1, ok) | Acc]);
+            OrderedEvents = lists:reverse(Events),
```
Comment (hanssv) [inline apps/aecore/src/aec_dry_run.erl:104]: Wouldn't it make sense to always deliver these in the correct order from `aec_trees`?

File: apps/aecore/src/aec_dry_run.erl [hunk 9]
```diff
@@ -96,20 +96,23 @@ dry_run_int([], _Trees, Env, _Opts, Acc) ->
 dry_run_int([{tx, TxOpts, Tx} | Txs], Trees, Env, Opts, Acc) ->
     Stateless = proplists:get_value(stateless, TxOpts, false),
     Env1 = prepare_env(Env, TxOpts),
+    EventsEnabled = proplists:get_bool(tx_events, Opts),
     %% GH3283: Here we should collect and present the `internal_call_tx` events.
     %% This means expanding the return type, and breaking the api :scream_cat:.
```
Comment (hanssv) [inline apps/aecore/src/aec_dry_run.erl:101]: Possibly remove this comment, since we do this, and break the api 😛

**Review by uwiger — DISMISSED** [review]

File: apps/aehttp/src/aehttp_app.erl (hunk 1 above)
Comment (seanhinde) [reply apps/aehttp/src/aehttp_app.erl:15]: Just following the existing convention. Re-doing the whole config thing is definitely needed, but I prefer to not do it here

File: apps/aeapi/src/aeapi.app.src [hunk 10]
```diff
@@ -0,0 +1,15 @@
+{application, aeapi,
+ [{description, "Common API layer"},
+  {vsn, {cmd, "cat ../../VERSION"}},
+  {registered, []},
+  {applications,
+   [kernel,
+    stdlib
```
[the remark this answers is not in the document]
Comment (seanhinde) [reply apps/aeapi/src/aeapi.app.src]: I added the main app it depends on (aecore)

File: apps/aecore/src/aec_dry_run.erl (hunk 8 above)
Comment (seanhinde) [reply apps/aecore/src/aec_dry_run.erl]: Yep. Done.

File: apps/aecore/src/aec_dry_run.erl (hunk 9 above)
Comment (seanhinde) [reply apps/aecore/src/aec_dry_run.erl]: Comment removed

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl (hunk 4 above)
Comment (seanhinde) [reply apps/aehttp/src/aehttp_dispatch_rosetta.erl:289]: As discussed will leave this as is for the moment. It's not performance critical and is clear enough

File: apps/aehttp/src/aehttp_dispatch_rosetta.erl (hunk 5 above)
Comment (seanhinde) [reply apps/aehttp/src/aehttp_dispatch_rosetta.erl:661]: As discussed will leave this as is for the moment. It's not performance critical and is clear enough

**Review by uwiger — APPROVED** [review]

**Review by hanssv — APPROVED** [review]

**Pull request merged**